ompi/mca/coll/acoll/coll_acoll_allreduce.c

```diff
@@ -0,0 +1,561 @@
+                             (char *) tmp_rbuf + chunk * rank * dsize, my_count_size, dtype);
+    }
+
+    err = ompi_coll_base_barrier_intra_tree(comm, module);
+    if (err != MPI_SUCCESS) {
+        return err;
+    }
+
+    for (int i = 1; i < size; i++) {
+        if (rank == i) {
+            continue;
+        }
+        ompi_op_reduce(op, (char *) data->xpmem_saddr[i] + chunk * rank * dsize,
+                       (char *) tmp_rbuf + chunk * rank * dsize, my_count_size, dtype);
+    }
+    err = ompi_coll_base_barrier_intra_tree(comm, module);
+    if (err != MPI_SUCCESS) {
+        return err;
+    }
+
+    size_t tmp = chunk * dsize;
+    for (int i = 0; i < size; i++) {
+        if (subc->xpmem_use_sr_buf && (rank == i)) {
+            continue;
+        }
+        my_count_size = (i == (size - 1)) ? (count / size) + count % size : count / size;
+        size_t tmp1 = i * tmp;
+        char *dst = (char *) rbuf + tmp1;
+        char *src = (char *) data->xpmem_raddr[i] + tmp1;
+        memcpy(dst, src, my_count_size * dsize);
+    }
+
+    err = ompi_coll_base_barrier_intra_tree(comm, module);
+
+    return err;
+}
+#endif
+
+void mca_coll_acoll_sync(coll_acoll_data_t *data, int offset, int *group, int gp_size, int rank,
+                         int up)
+{
+    volatile int *tmp, tmp0;
+    tmp = (int *) ((char *) data->allshmmmap_sbuf[group[0]] + offset
+                   + CACHE_LINE_SIZE * rank);
+    tmp0 = __atomic_load_n((int *) ((char *) data->allshmmmap_sbuf[group[0]] + offset
+                                    + CACHE_LINE_SIZE * group[0]),
+                           __ATOMIC_RELAXED);
+
+    opal_atomic_wmb();
+
+    int val;
+    if (up == 1) {
+        val = data->sync[0];
+    } else {
+        val = data->sync[1];
+    }
+
+    if (rank == group[0]) {
+        __atomic_store_n((int *) ((char *) data->allshmmmap_sbuf[group[0]] + offset
+                                  + CACHE_LINE_SIZE * group[0]),
+                         val, __ATOMIC_RELAXED);
+    }
+
+    while (tmp0 != val) {
+        tmp0 = __atomic_load_n((int *) ((char *) data->allshmmmap_sbuf[group[0]] + offset
+                                        + CACHE_LINE_SIZE * group[0]),
+                               __ATOMIC_RELAXED);
+    }
+
+    if (rank != group[0]) {
+        val++;
+        __atomic_store_n(tmp, val, __ATOMIC_RELAXED);
+    }
+    opal_atomic_wmb();
+    if (rank == group[0]) {
+        for (int i = 1; i < gp_size; i++) {
+            volatile int tmp1 = __atomic_load_n(
+                (int *) ((char *) data->allshmmmap_sbuf[group[0]] + offset + CACHE_LINE_SIZE * group[i]),
+                __ATOMIC_RELAXED);
+            while (tmp1 == val) {
+                tmp1 = __atomic_load_n((int *) ((char *) data->allshmmmap_sbuf[group[0]] + offset
+                                                + CACHE_LINE_SIZE * group[i]),
+                                       __ATOMIC_RELAXED);
+            }
```

Comment:
No need for `volatile` (I believe) and maybe we can use the `opal_atomic` API instead?

Reply:
We are planning to refactor some of the code in the next iteration, will keep your comment in mind. Just curious, what would be the benefit of opal_atomic?

Reply:
@devreal AFAIK we don't have `opal_atomic_load`/`opal_atomic_store` right? (related: #9722). We should at some point add them though :-), xhc will also make use of them.